### PR TITLE
Create OpenAPI Schema for basic use cases

### PR DESCRIPTION
**Pull request by lschlesinger**
# Create OpenAPI Schema for basic use cases
Detailed information on the OpenAPI Specification (OAS) to be found [here](https://swagger.io/specification/).

## :recycle: Current situation
As of now, no code related to the creation of OAS is in develop.

## :bulb: Proposed solution
The most relevant parts of the OAS were addressed first, leading to a `document` that contains:
- **`info`**: Rudimentary information in `info`
- **`server`**: Server address
- **`paths`**: Relative paths to the individual endpoints (i.e. each `Handler`), their operations, parameters (query and path), requestBody and responses (specified per HTTP status code)
- **`components/schemas`**: Definition of JSON schemas for objects (in contrast to primitive types) that can be referenced within the `paths - requestBody` or `paths - responses` section

As an example:
```swift
    struct SomeStruct: Vapor.Content {
        var someProp: Int
        var optionalInt: Int?
        var optionalString: String?
        var reqDouble: Double
    }

    struct SomeComp: Handler {
        @Parameter
        var name: String?

        func handle() -> SomeStruct {
            SomeStruct(someProp: 4, reqDouble: 5.0)
        }
    }
    
    struct User: Codable {
        var id: Int
        var name: String?
    }

    struct UserHandler: Handler {
        @Parameter var userId: Int

        func handle() -> User {
            User(id: userId)
        }
    }

    @PathParameter var userId: Int
    
    var content: some Component {
        Group("complexHandler") {
            SomeComp()
        }
        Group("user", $userId) {
            UserHandler(userId: $userId)
                    .operation(.update)
        }
    }
}
```

Will provide the following (for demonstration purposes somewhat condensed) OAS as output for the `document` sections `paths` and `components`:

```json
  "paths": {
    "/v1/user/{userId}": {
      "put": {
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/User"
                }
              }
            },
            "description": "OK"
          },
        "parameters": [
          {
            "description": "@Parameter var userId: Int",
            "in": "path",
            "name": "userId",
            "required": true,
            "schema": {
              "type": "integer"
            }
          }
        ]
      }
    },
    "/v1/complexHandler": {
      "get": {
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/SomeStruct"
                }
              }
            },
            "description": "OK"
          },
        "parameters": [
          {
            "in": "query",
            "name": "name",
            "schema": {
              "type": "string"
            },
            "description": "@Parameter var name: String?"
          }
        ]
      }
    }
  },
  "components": {
    "schemas": {
      "SomeStruct": {
        "type": "object",
        "properties": {
          "optionalInt": {
            "type": "integer"
          },
          "reqDouble": {
            "type": "number",
            "format": "double"
          },
          "optionalString": {
            "type": "string"
          },
          "someProp": {
            "type": "integer"
          }
        },
        "required": [
          "reqDouble",
          "someProp"
        ]
      },
      "User": {
        "type": "object",
        "properties": {
          "id": {
            "type": "integer"
          },
          "name": {
            "type": "string"
          }
        },
        "required": [
          "id"
        ]
      }
    }
  }
```

#### Provide functionality
OpenAPI exporter allows to access and further use an automatically created OAS for a webservice, i.e. RESTful application, described with Apodini DSL.

#### Utilize synergy effects

1. Provide data structures (in `Model/TypeInference` folder) for (recursive) type information retrieval (using `Runtime` lib) to be used in several exporter use cases (tackles issue #76)

As developed by @nityanandaz (#33), types can be recursively reflected and stored in a tree-like structure of nodes of type `EnrichedInfo`. 
The value of a `Node<EnrichedInfo>` is `EnrichedInfo` and `children` holds an array of `Node<EnrichedInfo>` for each property in `properties` of a reflected type.
`EnrichedInfo` holds fields of `TypeInfo` (`Runtime` type), `PropertyInfo` (`Runtime` type),  and `Cardinality` (information indicating optional or array types) and provides information on the `propertyOffset`.

The `OpenAPIExporter` exploits this data structure for reflecting the type of a request body (`EndpointParameter` with 
`parameterType` = `.content`) and the ultimate response of a `Handler` (stored in `responseType` of exported `Endpoint`).

2. Exploit information available in exported endpoint (already enriched on the level of the `SharedSemanticModelBuilder`)

E.g., Retrieve information associated to `Endpoint`, e.g. the determination of the ultimate `responseType` and the inference of an `EndpointParameter` s `parameterType`.

**Note**
The current version does not provide full support of corner cases and is to be further enhanced through future PRs.

#### Practical Implication
After starting the application, a JSON representation of the creates OAS is available under ``http://0.0.0.0:8080/openapi`` and can be used to generate client-side API in language of choice, e.g. using [swagger.io](https://editor.swagger.io/)

#### Refactoring and Sharing
- Models defined in #33 are lifted up to `Semantic Model Builder/Model/TypeInference` -> Proposal for refactoring the code related to `ProtobufferBuilder` done by 553f35e
- Other exporters that rely on recursively reflected type information can exploit the provided (tree, node) data structure and the associated operations (`map`, `filter`, `reduce`, etc.)

## :heavy_plus_sign: Additional Information

There are some corner cases that will be addressed in future PRs:
- optional array type: `let friends: [User]?` where the `Cardinality` cannot be reflected by _one_ enum typ but rather an array of cardinalities such as `[.oneToMany, .zeroToOne]` reflecting the cardinality of both, an optional and an array type.

- type definition including circuits: 
```swift
struct User {
var id: Int
var name: String
var friends: [User]
}
```

- Improve type inference for e.g. tuples, enums, dictionaries

- Support YAML export and improve how specification is served (unify with other exporters, e.g. for serve the `.proto` file)

### Related PRs
- The included models for type inference were developed by @nityanandaz and originate from his PR #33 

### Testing

 **`OpenAPIDocumentBuilderTests`**
- testing `OpenAPI.Document` for one `mockEndpoint` including the section `info`, `servers`, `paths` (including `query` parameter) and `components`

**`OpenAPIPathsObjectBuilderTests`**
- testing `OpenAPI.Path` built by `OpenAPIPathBuilder` including a path parameter
- testing `pathsObject` built by `OpenAPIPathsObjectBuilder` including query and path parameter for `.get` operation
- testing `pathsObject` built by `OpenAPIPathsObjectBuilder` including complex request body to be referenced in `components` and response type to be referenced in `components`
- testing `pathsObject` built by `OpenAPIPathsObjectBuilder` including two params wrapping complex request bodies that have to be wrapped by an artificial wrapper struct, include testing of `OpenAPIComponentsObjectBuilder` to reference this artificial wrapper struct in `components`

**`OpenAPIComponentsObjectBuilderTests`**
- testing build `OpenAPI.JSONSchema` built by `OpenAPIComponentsObjectBuilder` for primitive and complex types with (nested) references

### Reviewer Nudging
The "exporting workflow" is the following:
The `OpenAPIInterfaceExporter` initializes an `OpenAPIConfiguration` object (for general `info` and `server` section, retrieved from `app`) and creates a `OpenAPIDocumentBuilder` with this configuration on init.
On every export of an `Endpoint`, new information is added to the `paths` section by using the `OpenAPIComponentsObjectBuilder` and `OpenAPIComponentsObjectBuilder` which allows for referencing reusable types (defined in `components`) in the `paths` section. 

Changing the structure and content of `Handlers` in the main and afterwards investigating the served specification under `http://0.0.0.0:8080/openapi` will also provides insights on if/how corner cases are covered.

### CODE REVIEW

**Comment by bauer-andreas** [comment]
How is the status on this Draft? I saw you recently pulled in the latest changes from the parameter retrieval #81. Do you need any assistance on that to adjust the OpenAPI exporter?
I would love to see an initial prototype of the OpenAPI exporter getting merged into the develop branch. It doesn't have to be finished, continuous improvements can be made using additional PRs.
I think having the initial version of an exporter already in develop would make bigger refactorings like the parameter retrieval process easier and generally improves the overview for others (and review of smaller PRs are more pleasant 🙃).
Don't want to stress, its ready if you feel that it is ready 😊

**Comment by lschlesinger** [comment]
@nityanandaz I proposed a refactoring of your code in 553f35e - I am not sure if this is the way you've intended the structure. Feel free to suggest any changes. The reason for doing the refactoring was, that with the previous separation the OpenAPI and the ProtobufferBuilder code in different targets, I could not reach the status _one_ shared data structure representation in `Semantic Model Builder/Model/TypeInference`.

File: Sources/Apodini/SyntaxTreeVisitor/ContextNode.swift [hunk 1]
```diff
@@ -17,7 +17,7 @@ class ContextNode {
     }
     
     func copy() -> ContextNode {
-        let newContextNode = ContextNode(nodeLink: parentContextNode)
+        let newContextNode = ContextNode(nodeLink: parentContextNode?.copy())
```
Comment (bauer-andreas) [inline Sources/Apodini/SyntaxTreeVisitor/ContextNode.swift]: Just out of interest: Any particular reason why a complete copy here is necessary. Did I miss anything in the review that the OpenAPI exporter needs this?

File: Sources/Apodini/Semantic Model Builder/OpenAPI/OpenAPIInterfaceExporter.swift [hunk 2]
```diff
@@ -0,0 +1,94 @@
+//
+//  Created by Paul Schmiedmayer on 11/3/20.
+//
+
+@_implementationOnly import OpenAPIKit
+@_implementationOnly import Vapor
+import Foundation
+
+class OpenAPIInterfaceExporter: InterfaceExporter {
+    typealias ExporterRequest = Vapor.Request
```
Comment (bauer-andreas) [inline Sources/Apodini/Semantic Model Builder/OpenAPI/OpenAPIInterfaceExporter.swift]: Nit: `ExporterRequest` is defined as the argument in `retrieveParameter`. Using that definition of the function signature, swift should be able to automatically infer the type for `ExporterRequest`.

File: Sources/Apodini/Semantic Model Builder/OpenAPI/OpenAPIInterfaceExporter.swift [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                }
+            case .YAML:
+                print("Not implemented yet.")
+            default:
+                print("Not implemented yet.")
+            }
+        }
+    }
+
+    func retrieveParameter<Type: Decodable>(_ parameter: EndpointParameter<Type>, for request: Vapor.Request) throws -> Type?? {
```
Comment (bauer-andreas) [inline Sources/Apodini/Semantic Model Builder/OpenAPI/OpenAPIInterfaceExporter.swift:45]: The retrieveParameter function is copied from the REST Exporter right?
As OpenAPI is somewhat "static" exporter, meaning it doesn't actually expose any of its Endpoints. E.g. `handle()` won't ever be called, so `retrieveParameter` won't ever be executed for the OpenAPI exporter.

I see that such "static" exporters (e.g. which only serve information about the WebService but don't actually server the WebService itself) aren't properly supported right now. Honestly I completely forgot about that use case when reworking the Exporter API in #81. I see what I can do to properly support such scenarios.

Don't want to block you PR, so in the mean time I would propose to just replace the function body with a simple `fatalError` statement or something.

File: Sources/TestWebService/main.swift [hunk 4]
```diff
@@ -5,65 +5,32 @@
 //  Created by Paul Schmiedmayer on 7/6/20.
 //
 
-import Apodini
+@testable import Apodini
```
Comment (bauer-andreas) [inline Sources/TestWebService/main.swift]: I think Paul deliberately removed the `@testable` recently. Is there a particular reason why you added it back again, or did it just get lost while merging?

File: Sources/TestWebService/main.swift [hunk 5]
```diff
@@ -75,29 +42,14 @@ struct TestWebService: Apodini.WebService {
...
-            .response(EmojiMediator())
-            .guard(PrintGuard())
-        Group("swift") {
-            Text("Hello Swift! 💻")
-                .response(EmojiMediator())
-                .guard(PrintGuard())
-            Group("5", "3") {
-                Text("Hello Swift 5! 💻")
-            }
-        }.guard(PrintGuard("Someone is accessing Swift 😎!!"))
```
Comment (bauer-andreas) [inline Sources/TestWebService/main.swift]: why removing all these great Emoji endpoints 😢

Jokes. I think the current TestWebService with all is guards and response modifiers is a great way for us all to boot up a quick test environment. Any reason for removing those?

**Review by bauer-andreas — APPROVED** [review]
Great work 👍 

Added some questions at places which caught my eye.
In regards to `retrieveParameter` I would propose to replace it with an empty implementation, as pointed out in my comment below. Once this PR is merged I will rework the Exporter API to better support such types of Exporters.

Keep up the good work 🎉

File: Sources/Apodini/Semantic Model Builder/OpenAPI/OpenAPIInterfaceExporter.swift (hunk 2 above)
Comment (lschlesinger) [reply Sources/Apodini/Semantic Model Builder/OpenAPI/OpenAPIInterfaceExporter.swift]: resolved by 5693192

File: Sources/Apodini/Semantic Model Builder/OpenAPI/OpenAPIInterfaceExporter.swift (hunk 3 above)
Comment (lschlesinger) [reply Sources/Apodini/Semantic Model Builder/OpenAPI/OpenAPIInterfaceExporter.swift:45]: you're right, just copied :D replaced by fatalError in 5693192.

File: Sources/Apodini/SyntaxTreeVisitor/ContextNode.swift (hunk 1 above)
Comment (lschlesinger) [reply Sources/Apodini/SyntaxTreeVisitor/ContextNode.swift]: It's a left over from experimentation in earlier times 🦖 ... resolved by 73a2b4f

File: Sources/TestWebService/main.swift (hunk 5 above)
Comment (lschlesinger) [reply Sources/TestWebService/main.swift]: I shouldn't have added the `main.swift` to the PR. I have overwritten my version when merging the develop d784c8d - great Emoji endpoints are back 💯 🦜 🥳 🎈 🎉 🪂 🐘 🥨 🌵 🚗

**Comment by lschlesinger** [comment]
@nityanandaz please note that in e23a332 I had to rename your `Message` type to `ProtoMessage` as there was a naming conflict with the type that was introduced in the WebSocketExporter. With the restructuring in 553f35e, making ProtobufferBuilder not a separat package anymore, we had `Message` for both use cases in the same scope. I hope this is a reasonable refactoring, couldn't come up with a better solution.

File: Sources/TestWebService/main.swift (hunk 4 above)
Comment (lschlesinger) [reply Sources/TestWebService/main.swift]: resolved by d784c8d

**Review by bauer-andreas — APPROVED** [review]

File: Sources/Apodini/Semantic Model Builder/gRPC/ProtobufferBuilder/Model.swift [hunk 6]
```diff
@@ -7,7 +7,7 @@
 
 // MARK: - Message
 
-struct Message: Equatable, Hashable {
+struct ProtoMessage: Equatable, Hashable {
```
Comment (nityanandaz) [inline Sources/Apodini/Semantic Model Builder/gRPC/ProtobufferBuilder/Model.swift]: @moritzschuell and I agreed on not using abbreviations here: https://github.com/Apodini/Apodini/pull/43#discussion_r549125372

Please rename Service as well if this change is necessary.

**Review by nityanandaz — CHANGES_REQUESTED** [review]

File: Sources/Apodini/Semantic Model Builder/gRPC/ProtobufferBuilder/Model.swift (hunk 6 above)
Comment (lschlesinger) [reply Sources/Apodini/Semantic Model Builder/gRPC/ProtobufferBuilder/Model.swift]: solved by 57e6975

File: Sources/Apodini/Semantic Model Builder/OpenAPI/Document/OpenAPIComponentsObjectBuilder.swift [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    func buildSchema(for type: Encodable.Type) throws -> JSONSchema {
+        let node: Node<JSONSchema>? = try Self.node(type)?
+                .contextMap(contextMapNode)
+        guard let schema = node?.value else {
+            throw OpenAPIComponentBuilderError("Could not reflect type.")
+        }
+        return schema
+    }
+
+    private func contextMapNode(node: Node<EnrichedInfo>) -> JSONSchema {
```
Comment (nityanandaz) [inline Sources/Apodini/Semantic Model Builder/OpenAPI/Document/OpenAPIComponentsObjectBuilder.swift:62]: I assume, we shouldn't track issues in code. I referenced your issue here: https://github.com/Apodini/Apodini/issues/95

File: Sources/Apodini/Semantic Model Builder/OpenAPI/Document/OpenAPIPathsObjectBuilder.swift [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        // Get `OpenAPI.HttpMethod` and `OpenAPI.Operation` from endpoint.
+        let httpMethod = endpoint.operation.openAPIHttpMethod
+        let operation = buildPathItemOperationObject(from: endpoint)
+        pathItem.set(operation: operation, for: httpMethod)
+
+        // Add (or override) `PathItem` to map of paths.
+        pathsObject[path] = pathItem
+    }
+
+
```
Comment (nityanandaz) [inline Sources/Apodini/Semantic Model Builder/OpenAPI/Document/OpenAPIPathsObjectBuilder.swift]: Same as here: https://github.com/Apodini/Apodini/pull/90#discussion_r551261669

😄

File: Sources/Apodini/Semantic Model Builder/OpenAPI/Extensions/AnyEndpointParameter+OpenAPI.swift [hunk 9]
```diff
@@ -0,0 +1,20 @@
+//
+//  Created by Lorena Schlesinger on 09.12.20.
+//
+
+import Foundation
+@_implementationOnly import OpenAPIKit
+
```
Comment (nityanandaz) [inline Sources/Apodini/Semantic Model Builder/OpenAPI/Extensions/AnyEndpointParameter+OpenAPI.swift]:
```suggestion
extension OpenAPI.Parameter.Context {
    init?(_ endpointParameter: AnyEndpointParameter) {
        switch endpointParameter.parameterType {
        case .lightweight:
            self = .query
        case .content:
            return nil
        case .path:
            self = .path
        }
    }
}
```

This is just a suggestion, but I worry, that core types of Apodini can get polluted with extensions that are built on top of them and don't reflect their inherent functionality.

This approach should deflect the growing duty on these core Apodini types.

What do you think?

File: Sources/Apodini/Semantic Model Builder/OpenAPI/OpenAPIConfiguration.swift [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    enum OutputFormat {
+        case JSON
+        case YAML
+    }
+
+    var outputPath: String?
+    var outputEndpoint: String? = "openapi"
+    var outputFormat: OutputFormat = .JSON
+
+    static func create(from app: Application) -> Self {
```
Comment (nityanandaz) [inline Sources/Apodini/Semantic Model Builder/OpenAPI/OpenAPIConfiguration.swift]: Synthesized initializers (for structs) are preserved if alternative initializers are declared in an extension. I suggest:

```
extension OpenAPIConfiguration {
    init(app: Application) {
        let host = app.http.server.configuration.hostname
        let port = app.http.server.configuration.port
        var servers: [OpenAPI.Server] = []
        if let url = URL(string: "\(host):\(port)") {
            let server = OpenAPI.Server(url: url)
            servers.append(server)
        }
        
        self.init(servers: servers)
    }
}
```

File: Sources/Apodini/Semantic Model Builder/OpenAPI/OpenAPIInterfaceExporter.swift [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    func export<H: Handler>(_ endpoint: Endpoint<H>) {
+        documentBuilder.addEndpoint(endpoint)
+    }
+
+    func finishedExporting(_ webService: WebServiceModel) {
+        serveSpecification()
+    }
+
+    private func serveSpecification() {
+        // swiftlint:disable:next todo
```
Comment (nityanandaz) [inline Sources/Apodini/Semantic Model Builder/OpenAPI/OpenAPIInterfaceExporter.swift]: Again, I'd rather not have TODOs in code. Why not create an issue 🙂

File: Tests/ApodiniTests/OpenAPITests/OpenAPIComponentsObjectBuilderTests.swift [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+    struct SomeStruct: Encodable {
+        var id: UUID?
+        var someProp = 4
+    }
+
+    struct GenericStruct<T>: Encodable {
+        var list: [T]
+        var listLength: Int
+
+        func encode(to encoder: Encoder) throws {
```
Comment (nityanandaz) [inline Tests/ApodiniTests/OpenAPITests/OpenAPIComponentsObjectBuilderTests.swift]: `struct GenericStruct<T>: Encodable where T: Encodable { ... }` should do the trick.

**Review by nityanandaz — CHANGES_REQUESTED** [review]
Furthermore, I would rename `Sources/Apodini/Semantic Model Builder/Model/TypeInference`

I think, `TypeReflection` or `TypeInspection` is more fitting. Type inference is a different activity 🙂

Ref.: https://en.wikipedia.org/wiki/Type_inference

File: Sources/Apodini/Semantic Model Builder/OpenAPI/OpenAPIInterfaceExporter.swift (hunk 11 above)
Comment (lschlesinger) [reply Sources/Apodini/Semantic Model Builder/OpenAPI/OpenAPIInterfaceExporter.swift]: #107

File: Tests/ApodiniTests/OpenAPITests/OpenAPIComponentsObjectBuilderTests.swift (hunk 12 above)
Comment (lschlesinger) [reply Tests/ApodiniTests/OpenAPITests/OpenAPIComponentsObjectBuilderTests.swift]: done in 6974958

File: Sources/Apodini/Semantic Model Builder/OpenAPI/OpenAPIConfiguration.swift (hunk 10 above)
Comment (lschlesinger) [reply Sources/Apodini/Semantic Model Builder/OpenAPI/OpenAPIConfiguration.swift]: good point! thanks, done in 650ebc2

File: Sources/Apodini/Semantic Model Builder/OpenAPI/Extensions/AnyEndpointParameter+OpenAPI.swift (hunk 9 above)
Comment (lschlesinger) [reply Sources/Apodini/Semantic Model Builder/OpenAPI/Extensions/AnyEndpointParameter+OpenAPI.swift]: Absolutetly right, thanks for the suggestion!! Refactored in 5e54667

File: Sources/Apodini/Semantic Model Builder/OpenAPI/Document/OpenAPIPathsObjectBuilder.swift (hunk 8 above)
Comment (lschlesinger) [reply Sources/Apodini/Semantic Model Builder/OpenAPI/Document/OpenAPIPathsObjectBuilder.swift]: nice idea! done in c67bfda

File: Sources/Apodini/Semantic Model Builder/OpenAPI/Extensions/Operation+OpenAPI.swift [hunk 13]
```diff
@@ -0,0 +1,24 @@
+//
+//  Created by Lorena Schlesinger on 15.11.20.
+//
+
+import Foundation
+@_implementationOnly import OpenAPIKit
+
+/// Extension to map Apodini `Operation`  to `OpenAPI.HttpMethod`.
+extension Operation {
```
Comment (nityanandaz) [inline Sources/Apodini/Semantic Model Builder/OpenAPI/Extensions/Operation+OpenAPI.swift]: I'd follow the same approach here as well (https://github.com/Apodini/Apodini/pull/35#discussion_r552936299)

```suggestion
extension OpenAPI.HttpMethod {
    init(_ operation: Operation) {
        switch operation {
        case .automatic, .read:
            self = .get
        case .create:
            self = .post
        case .update:
            self = .put
        case .delete:
            self = .delete
        }
    }
}
```

**Review by nityanandaz — APPROVED** [review]

File: Sources/Apodini/Semantic Model Builder/OpenAPI/Extensions/Operation+OpenAPI.swift (hunk 13 above)
Comment (lschlesinger) [reply Sources/Apodini/Semantic Model Builder/OpenAPI/Extensions/Operation+OpenAPI.swift]: solved by 5b47895

**Pull request merged**